### DISCUSSION TITLE
Make Frame clonable

### DISCUSSION THREAD

**Discussion by Sytten** [hyperium/hyper#3768] [General]
Hi!

I am aware that the change to a `Frame` enum was made to allow users to give their own `Frameish` implementation.
Though this has not been done yet, I assume it is still relevant and will be done eventually.

I would like to propose that `Frame` should implement `Clone` and thus that the future `Frameish` trait also be bounded by `Clone`. This allows the consumer of a body to broadcast those frames to multiple consumers without having to aggregate all the body first. This would not be a breaking change right now, that is why I am proposing it before the trait is implemented.

I do get that it is always better to not introduce this restriction since we don't really know what use-cases might be for `Frameish`, but at the same time I don't think it is that much of a bar since worst case the users can wrap their frame decoder in `Bytes` to make them clonable.

**Comment by seanmonstar** [comment]
Thanks for the writeup, you tackled both side well!

As you noticed, the frame design was made for forwards-compatibility in mind. Not with a specific immediate need, but to allow for new frames from h2 or h3. Because that was its entire purpose, I lean most towards keeping it maximally flexible.

I do appreciate that makes it harder to tee the stream...

**Reply by Sytten** [reply]
Yeah basically we are implementing a tee for quickjs engine in llrt (https://github.com/awslabs/llrt) and it's kinda a pain if we can't clone it. Right now I worked around it, but if you do introduce the boxed trait and it is not clonable then we won't be able to do that trick anymore and it will effectively impossible to tee (without unsafe). Per say it would only be `Clone` if T is `Clone`, but I understand that is limits `Frameish` to `Clone` too. Anyway at least it is on your radar of concerns.

```rust
#[derive(Clone)]
enum ClonableFrame<T> {
    Data(T),
    Trailers(HeaderMap),
}

impl<T> From<Frame<T>> for ClonableFrame<T> {
    fn from(frame: Frame<T>) -> Self {
        let frame = match frame.into_data() {
            Ok(data) => return ClonableFrame::Data(data),
            Err(frame) => frame,
        };

        match frame.into_trailers() {
            Ok(trailers) => ClonableFrame::Trailers(trailers),
            Err(_) => unreachable!(),
        }
    }
}

impl<T> From<ClonableFrame<T>> for Frame<T> {
    fn from(frame: ClonableFrame<T>) -> Self {
        match frame {
            ClonableFrame::Data(data) => Frame::data(data),
            ClonableFrame::Trailers(trailers) => Frame::trailers(trailers),
        }
    }
}
```
